### ISSUE TITLE
 OneSignal.appContext null, could not read OneSignal from getSharedPreferences.

### ISSUE THREAD

**Issue by vijayawoshe** [OneSignal/OneSignal-Flutter-SDK#177]
**Description:**
I am facing the following issue with version 2.2.0.
Could you please help me on this issue. its urgent.

```java
OneSignal.appContext null, could not read OneSignal from getSharedPreferences.
W/OneSignal(  966): java.lang.Throwable
W/OneSignal(  966): 	at com.onesignal.OneSignalPrefs.getSharedPrefsByName(OneSignalPrefs.java:279)
W/OneSignal(  966): 	at com.onesignal.OneSignalPrefs.get(OneSignalPrefs.java:255)
W/OneSignal(  966): 	at com.onesignal.OneSignalPrefs.getBool(OneSignalPrefs.java:223)
W/OneSignal(  966): 	at com.onesignal.OneSignal.getSavedUserConsentStatus(OneSignal.java:2170)
W/OneSignal(  966): 	at com.onesignal.OneSignal.userProvidedPrivacyConsent(OneSignal.java:769)
W/OneSignal(  966): 	at com.onesignal.OneSignal.requiresUserPrivacyConsent(OneSignal.java:1060)
W/OneSignal(  966): 	at com.onesignal.flutter.OneSignalPlugin.onMethodCall(OneSignalPlugin.java:92)
W/OneSignal(  966): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:231)
W/OneSignal(  966): 	at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(DartMessenger.java:93)
W/OneSignal(  966): 	at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(FlutterJNI.java:642)
W/OneSignal(  966): 	at android.os.MessageQueue.nativePollOnce(Native Method)
W/OneSignal(  966): 	at android.os.MessageQueue.next(MessageQueue.java:336)
W/OneSignal(  966): 	at android.os.Looper.loop(Looper.java:174)
W/OneSignal(  966): 	at android.app.ActivityThread.main(ActivityThread.java:7356)
W/OneSignal(  966): 	at java.lang.reflect.Method.invoke(Native Method)
W/OneSignal(  966): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
W/OneSignal(  966): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
I/flutter (  966): Check Connection...
E/OneSignal(  966): Cannot change requiresUserPrivacyConsent() from TRUE to FALSE
```

**Environment**
`onesignal_flutter: 2.2.0`

**Steps to Reproduce Issue:**

```dart
 static var settings = {
    OSiOSSettings.autoPrompt: false,
    OSiOSSettings.promptBeforeOpeningPushUrl: true,
    OSiOSSettings.inAppLaunchUrl: true
  };

  static initOneSignal() async {
    print("___________________INIT ONE SIGNAL________________________");
    await OneSignal.shared.setRequiresUserPrivacyConsent(true);
    await OneSignal.shared.init(ONESIGNAL_APP_ID, iOSSettings: settings);
    OneSignal.shared.setLogLevel(OSLogLevel.verbose, OSLogLevel.none);
```

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel unknown, v1.11.0, on Mac OS X 10.15.1 19B88, locale en-GB)
 
[!] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    ✗ Android license status unknown.
      Try re-installing or updating your Android SDK Manager.
      See https://developer.android.com/studio/#downloads or visit https://flutter.dev/setup/#android-setup for detailed instructions.
[✓] Xcode - develop for iOS and macOS (Xcode 11.2.1)
[✓] Android Studio (version 3.5)
[✓] VS Code (version 1.40.1)
[✓] Connected device (1 available)

```

**Comment by jkasten2** [comment]
@vijayawoshe Thanks for reporting.

The following you are seeing the in the log is a warning not an error.
`OneSignal.appContext null, could not read OneSignal from getSharedPreferences.`
It is something we can look into cleaning up in the future.

The following is an error in your code however.
`E/OneSignal(  966): Cannot change requiresUserPrivacyConsent() from TRUE to FALSE`

Can you search your app for other places where you are calling `OneSignal.shared.setRequiresUserPrivacyConsent`? It seems some where you are passing `false` to this somewhere.
Note, if the user accepts your privacy policy you should call `OneSignal.shared.consentGranted(true)`

**Comment by vijayawoshe** [comment]
Hello @jkasten2 ,

I checked in my application, I am not calling setRequiresUserPrivacyConsent with false anywhere. I am just calling this method with true on app init.

I dont know why I am still getting this message.

Could you please help?

Thanks

Regards
Vijay

**Comment by rgomezp** [comment]
Howdy Vijay,
Sorry for the delay. Were you able to resolve?

**Comment by enricobenedos** [comment]
I'm experiencing the same problem in a new project. The very strange thing is that if you use only an iOS device, this debug line is not printed.
```Cannot change requiresUserPrivacyConsent() from TRUE to FALSE```
This my `one signal` package init code:
```dart
    //Remove this method to stop OneSignal Debugging
    OneSignal.shared.setLogLevel(OSLogLevel.verbose, OSLogLevel.none);
    //Configuring OneSignal SDK
    OneSignal.shared.init("myAppId", iOSSettings: {
      OSiOSSettings.autoPrompt: false,
      OSiOSSettings.inAppLaunchUrl: false
    });
    OneSignal.shared
        .setInFocusDisplayType(OSNotificationDisplayType.notification);
    OneSignal.shared.setRequiresUserPrivacyConsent(true);
```
an this is the code I use to switch between notifications on or off:
```dart
onChanged: (bool value) async {
                    //Check if user has accepted privacy policy about notifications
                    final bool requiredPrivacy =
                        await OneSignal.shared.requiresUserPrivacyConsent();

                    if (requiredPrivacy) {
                      //TODO: improve dialog to include images and privacy policy
                      final result = await GlobalMethods.showChooseDialog(
                          context,
                          'Enable Notifications',
                          'Do you really want to enable notifications?');
                      if (result) {
                        await OneSignal.shared
                            .consentGranted(true);
                        if (Platform.isIOS) {
                          await OneSignal.shared
                              .promptUserForPushNotificationPermission(
                                  fallbackToSettings: true);
                        }
                      }
                    }

                    final bool requiredPrivacy2 =
                        await OneSignal.shared.requiresUserPrivacyConsent();
                    if (!requiredPrivacy2) {
                      await OneSignal.shared
                          .setSubscription(!_notificationsEnabled);

                      setState(() {
                        _notificationsEnabled = !_notificationsEnabled;
                      });
                    }
                  },
```

**Comment by Lawati97** [comment]
Same here it happens only if the user has not registered to One Signal before and it is the first time using the application

**Comment by Lawati97** [comment]
E[/OneSignal]()(16100): Cannot change requiresUserPrivacyConsent() from TRUE to FALSE
W[/OneSignal]()(16100): appContext set, but please call setAppId(appId) with a valid appId to complete OneSignal init!

**Comment by nan-li** [comment]
Hi all, sorry for the delayed response. 

This log is a logging mistake and can be safely ignored if it does not apply to you.
```
Cannot change requiresUserPrivacyConsent() from TRUE to FALSE
```

It has been fixed  to clean up this confusing log message in https://github.com/OneSignal/OneSignal-Android-SDK/pull/1527 and released with the [OneSignal-Flutter-SDK 3.3.0](https://github.com/OneSignal/OneSignal-Flutter-SDK/releases/tag/3.3.0).
